src/youtube-player/youtube-player.ts

```diff
@@ -0,0 +1,216 @@
+/// <reference types="youtube"/>
+
+import {
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  Component,
+  ElementRef,
+  EventEmitter,
+  Input,
+  OnDestroy,
+  ViewChild,
+} from '@angular/core';
+
+import {
+  combineLatest,
+  of as observableOf,
+  Observable,
+  ConnectableObservable,
+  pipe,
+  MonoTypeOperatorFunction,
+  merge,
+  OperatorFunction,
+} from 'rxjs';
+
+import {
+  map,
+  scan,
+  withLatestFrom,
+  flatMap,
+  filter,
+  startWith,
+  publish,
+  first,
+  distinctUntilChanged,
+  takeUntil,
+} from 'rxjs/operators';
+
+// The native YT.Player doesn't expose the set videoId, but we need it for
+// convenience.
+interface Player extends YT.Player {
+  videoId?: string | undefined;
+}
+
+// The player isn't fully initialized when it's constructed.
+// The only field available is destroy and addEventListener.
+type UninitializedPlayer = Pick<Player, 'videoId' | 'destroy' | 'addEventListener'>;
+
+/**
+ * Simple component that previews a YouTube video.
+ */
+@Component({
+  selector: 'youtube-player',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  templateUrl: 'youtube-player.html',
+})
+export class YouTubePlayer implements AfterViewInit, OnDestroy {
+  /** YouTube Video ID to view */
+  @Input() set videoId(videoId: string | undefined) {
+    this._videoId.emit(videoId);
+  }
+  private _videoId = new EventEmitter<string | undefined>();
+
+  /** The element that will be replaced by the iframe. */
+  @ViewChild('youtube_container', {static: false}) youtubeContainer: ElementRef | undefined;
+  private _youtubeContainer = new EventEmitter<HTMLElement>();
+  private _destroyed = new EventEmitter<undefined>();
+
+  private _player: Player | undefined;
+
+  constructor() {
+    const playerObs =
+      createPlayerObservable(
```

Comment:
Can you add a high-level description of everything captured by this stream here?